// nestjs-slonik/dist/slonik-core.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var SlonikCoreModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlonikCoreModule = void 0;
const common_1 = require("@nestjs/common");
const core_1 = require("@nestjs/core");
const rxjs_1 = require("rxjs");
const slonik_1 = require("slonik");
const common_2 = require("./common");
const slonik_constants_1 = require("./slonik.constants");
let SlonikCoreModule = SlonikCoreModule_1 = class SlonikCoreModule {
    constructor(options, moduleRef) {
        this.options = options;
        this.moduleRef = moduleRef;
        this.logger = new common_1.Logger('SlonikModule');
    }
    static forRoot(options) {
        const slonikOptions = {
            provide: slonik_constants_1.SLONIK_MODULE_OPTIONS,
            useValue: options,
        };
        const poolProvider = {
            provide: common_2.getPoolToken(options),
            useFactory: () => __awaiter(this, void 0, void 0, function* () { return this.createPoolFactory(options); }),
        };
        return {
            module: SlonikCoreModule_1,
            providers: [poolProvider, slonikOptions],
            exports: [poolProvider],
        };
    }
    static forRootAsync(options) {
        const poolProvider = {
            provide: common_2.getPoolToken(options),
            useFactory: (slonikOptions) => __awaiter(this, void 0, void 0, function* () {
                if (options.name) {
                    return this.createPoolFactory(Object.assign(Object.assign({}, slonikOptions), { name: options.name }));
                }
                return this.createPoolFactory(slonikOptions);
            }),
            inject: [slonik_constants_1.SLONIK_MODULE_OPTIONS],
        };
        const asyncProviders = this.createAsyncProviders(options);
        return {
            module: SlonikCoreModule_1,
            imports: options.imports,
            providers: [
                ...asyncProviders,
                poolProvider,
                {
                    provide: slonik_constants_1.SLONIK_MODULE_ID,
                    useValue: common_2.generateString(),
                },
            ],
            exports: [poolProvider],
        };
    }
    onApplicationShutdown() {
        return __awaiter(this, void 0, void 0, function* () {
            const pool = this.moduleRef.get(common_2.getPoolToken(this.options));
            try {
                yield (pool === null || pool === void 0 ? void 0 : pool.end());
            }
            catch (e) {
                this.logger.error(e === null || e === void 0 ? void 0 : e.message);
            }
        });
    }
    static createAsyncProviders(options) {
        if (options.useExisting || options.useFactory) {
            return [this.createAsyncOptionsProvider(options)];
        }
        const useClass = options.useClass;
        return [
            this.createAsyncOptionsProvider(options),
            {
                provide: useClass,
                useClass,
            },
        ];
    }
    static createAsyncOptionsProvider(options) {
        if (options.useFactory) {
            return {
                provide: slonik_constants_1.SLONIK_MODULE_OPTIONS,
                useFactory: options.useFactory,
                inject: options.inject || [],
            };
        }
        const inject = [
            (options.useClass || options.useExisting),
        ];
        return {
            provide: slonik_constants_1.SLONIK_MODULE_OPTIONS,
            useFactory: (optionsFactory) => __awaiter(this, void 0, void 0, function* () { return optionsFactory.createSlonikOptions(options.name); }),
            inject,
        };
    }
    static createPoolFactory(options) {
        return __awaiter(this, void 0, void 0, function* () {
            const poolToken = common_2.getPoolName(options);
            return rxjs_1.defer(() => __awaiter(this, void 0, void 0, function* () {
                const pool = yield slonik_1.createPool(options.connectionUri, options.clientConfigurationInput);
                yield pool.connect(() => Promise.resolve());
                return pool;
            }))
                .pipe(common_2.handleRetry(options.retryAttempts, options.retryDelay, poolToken, options.verboseRetryLog, options.toRetry))
                .toPromise();
        });
    }
};
SlonikCoreModule = SlonikCoreModule_1 = __decorate([
    common_1.Global(),
    common_1.Module({}),
    __param(0, common_1.Inject(slonik_constants_1.SLONIK_MODULE_OPTIONS)),
    __metadata("design:paramtypes", [Object, core_1.ModuleRef])
], SlonikCoreModule);
exports.SlonikCoreModule = SlonikCoreModule;
